### PR TITLE
Add validations for immutable AzureCluster Updates

### LINKED ISSUE #497: Add validations for AzureCluster Updates
Some fields (eg. Azure location) are immutable and we should fail when they are updated.

To get started, you can look at how we implemented validation for AzureMachines here: 

https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/api/v1alpha3/azuremachine_webhook.go#L46

and add to https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/api/v1alpha3/azurecluster_webhook.go

You can also look at this PR in CAPA which achieves something similar: https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/1677/files

### PR DESCRIPTION
**Pull request by furkatgofurov7**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Adds validations and corresponding unit tests for AzureCluster Updates
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #497 

**Special notes for your reviewer**:

**TODOs**:

- [X] squashed commits
- [ ] includes documentation
- [X] adds unit tests

**Release note**:

```release-note
Add validations for AzureCluster Updates
```

### CODE REVIEW

**Comment by furkatgofurov7** [comment]
@devigned @CecileRobertMichon I am trying to fix e2e tests and looking for some guidance. As I checked, webhook config for cluster is already in place, am I missing something else to pass the tests?

**Comment by devigned** [comment]
@furkatgofurov7 seems like something is up with kind. I'm, looking :eyes:

**Comment by furkatgofurov7** [comment]
@devigned hi! I was wondering if I am missing something to pass e2e tests or whether this is a flake in CI(most probably not), since we already have triggered it multiple times..?

**Comment by devigned** [comment]
@furkatgofurov7 I think the validating webhook is rejecting updates to the `AzureCluster` resource as evidenced by the repeated updates here: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/1098/pull-cluster-api-provider-azure-e2e/1346080959979589632/artifacts/clusters/bootstrap/controllers/capz-controller-manager/capz-controller-manager-5d7858d64d-nc2p8/manager.log

If you click on the e2e job, you can see the output of most of the cluster logs in the PR Artifacts link.

Have you been able to verify these changes locally by creating a cluster, perhaps using Tilt?

File: api/v1alpha3/azurecluster_webhook.go [hunk 1]
```diff
@@ -56,8 +60,37 @@ func (c *AzureCluster) ValidateCreate() error {
...
+	}
+
+	if !reflect.DeepEqual(c.Spec.Location, old.Spec.Location) {
+		allErrs = append(allErrs,
+			field.Invalid(field.NewPath("spec", "Location"),
+				c.Spec.Location, "field is immutable"),
+		)
+	}
+
+	if !reflect.DeepEqual(c.Spec.IdentityRef, old.Spec.IdentityRef) {
```
Comment (nader-ziada) [inline api/v1alpha3/azurecluster_webhook.go:87]: I don't think the identityRef field should be immutable, @devigned wdyt?

Comment (devigned) [reply api/v1alpha3/azurecluster_webhook.go:87]: Definitely not immutable.

Comment (CecileRobertMichon) [reply api/v1alpha3/azurecluster_webhook.go:87]: Yeah, I didn't put it in the list in #497 because it wasn't there at the time. Should be mutable.

Comment (furkatgofurov7) [reply api/v1alpha3/azurecluster_webhook.go]: Removed, I was not sure of it so identityRef was added to the list of immutable fields here. Thanks for review!

File: api/v1alpha3/azurecluster_webhook.go [hunk 2]
```diff
@@ -56,8 +60,30 @@ func (c *AzureCluster) ValidateCreate() error {
 // ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
 func (c *AzureCluster) ValidateUpdate(oldRaw runtime.Object) error {
 	clusterlog.Info("validate update", "name", c.Name)
+	var allErrs field.ErrorList
 	old := oldRaw.(*AzureCluster)
-	return c.validateCluster(old)
```
Comment (CecileRobertMichon) [inline api/v1alpha3/azurecluster_webhook.go:60]: we still want to run `validateCluster` to make sure that the fields that did get changed are still valid

File: api/v1alpha3/azurecluster_webhook.go [hunk 3]
```diff
@@ -56,8 +60,30 @@ func (c *AzureCluster) ValidateCreate() error {
...
+		)
+	}
+
+	if !reflect.DeepEqual(c.Spec.Location, old.Spec.Location) {
+		allErrs = append(allErrs,
+			field.Invalid(field.NewPath("spec", "Location"),
+				c.Spec.Location, "field is immutable"),
+		)
+	}
+	return apierrors.NewInvalid(GroupVersion.WithKind("AzureCluster").GroupKind(), c.Name, allErrs)
```
Comment (CecileRobertMichon) [inline api/v1alpha3/azurecluster_webhook.go:86]: I think the issue here (and why e2e is failing) is that this is always returning an error, even if allErrs is empty. It should return nil if len(allErrs) is 0.

Comment (furkatgofurov7) [reply api/v1alpha3/azurecluster_webhook.go:91]: Thanks, @CecileRobertMichon for review, forgot to add a check :). Updated.

**Comment by furkatgofurov7** [comment]
@CecileRobertMichon @devigned @nader-ziada thanks for reviews/feedbacks, addressed all your comments. PTAL.

File: api/v1alpha3/azurecluster_webhook.go [hunk 4]
```diff
@@ -56,8 +61,38 @@ func (c *AzureCluster) ValidateCreate() error {
...
+
+	if !reflect.DeepEqual(c.Spec.Location, old.Spec.Location) {
+		allErrs = append(allErrs,
+			field.Invalid(field.NewPath("spec", "Location"),
+				c.Spec.Location, "field is immutable"),
+		)
+	}
+
+	if len(allErrs) == 0 {
+		return nil
```
Comment (CecileRobertMichon) [inline api/v1alpha3/azurecluster_webhook.go:92]: What do you think about doing `return  c.validateCluster(old)` here if no immutable fields were changed? That way we're first checking if anything was changed that shouldn't have changed then validating the remaining fields.

Comment (CecileRobertMichon) [reply api/v1alpha3/azurecluster_webhook.go:92]: (instead of returning a new error for validateCluster above)

**Comment by CecileRobertMichon** [comment]
@furkatgofurov7 please add a release note as this is a user facing change

**Review by CecileRobertMichon — COMMENTED** [review]
lgtm

please squash the commits @furkatgofurov7

File: api/v1alpha3/azurecluster_webhook.go (hunk 4 above)
Comment (furkatgofurov7) [reply api/v1alpha3/azurecluster_webhook.go]: Done.

**Comment by furkatgofurov7** [comment]
> @furkatgofurov7 please add a release note as this is a user facing change

Done.


> please squash the commits @furkatgofurov7

Done. Thanks for review @CecileRobertMichon .

**Review by CecileRobertMichon — APPROVED** [review]
/lgtm
/approve
/retest

**Comment by CecileRobertMichon** [comment]
/retest

#1123 should help with the flakes...

**Pull request merged**